API: allow rechunk to evenly split into N chunks

stsievert opened the pull request:
**What does this PR implement?**
The ability to find the chunksize to evenly chunk an array. For example, say I want to split my array of length 1010 into about 20 chunks. What chunk size is best for that? And what if I want don't mind if I have 19 or 21 chunks?

**References issues/PRs**
This PR is motivated by https://github.com/dask/dask-blog/pull/66, which is running a comparison benchmark at https://github.com/stsievert/dask-hyperband-comparison/pull/3.

When following a modification of Dask-ML's `HyperbandSearchCV` docs, I have this code:

``` python
>>> n_params = 75
>>> n_epochs = 200
>>> n_chunks = n_epochs / n_params
```

I would like to chunk two NumPy training arrays of length 45,000 into `n_chunks` evenly sized chunks, or close to it. This PR provides the implementation to find the chunk size that's best (number of chunks close to `n_chunks`, and they're all equally sized):

``` python
>>> y_train = np.random.uniform(size=45_000)
>>> chunksize = da.even_chunksize(len(y_train), n_epochs / n_params)
>>> y_train2 = da.from_array(y_train, chunks=chunksize)
>>> y_train2.chunks
((22500, 22500, 15000),)
```

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

mrocklin commented:
Hrm, this is close enough to something like `x = x.rechunk()` that I would be inclined to see if something like that would work for your use case.  If we do go in this direction we'll have to figure out an API and semantics that make sense for multiple dimensions.  

Do you have a strong need to specify the exact number of chunks, or would something like `x = x.rechunk("100 MiB")` work?

stsievert commented:
> Do you have a strong need to specify the exact number of chunks?

Yes. I slightly reformulated [HyperbandSearchCV's rule-of-thumb][1]. Now, it recommends a floating point number of chunks; in the benchmark I'm running for https://github.com/dask/dask-blog/pull/66, `2.666` chunks are recommended. The new rule-of-thumb avoids the user modifying `HyperbandSearchCV`'s input parameters to make sure the input arrays chunk evenly.*

> this is close enough to something like x = x.rechunk()

I think `X.rechunk(n_chunks=(3, -1))` is a better API for a 2D array `X` to have 3 chunks in the 1st dimension and no chunks in the second dimension. I'll push that change.

<sup>* [The HyperbandSearchCV example][2] say "Approximately 300 parameters were desired; 299 was chosen to make the Dask array chunk evenly" in a footnote.</sup>

[2]:https://ml.dask.org/hyper-parameter-search.html#hyperband-performance
[1]:https://ml.dask.org/hyper-parameter-search.html#hyperband-parameters-rule-of-thumb

mrocklin commented on dask/array/rechunk.py:
```diff
@@ -220,9 +226,20 @@ def rechunk(x, chunks="auto", threshold=None, block_size_limit=None):
     >>> y = x.rechunk({0: -1, 1: 'auto'}, block_size_limit=1e8)
     """
+    if n_chunks:
+        if isinstance(n_chunks, (int, float)):
+            n_chunks = (n_chunks,) * x.ndim
```
I wonder if this behavior might surprise people in the tall-and-skinny case common in ML workflows.  I don't really have a better suggestion here, other than to force the user to provide a tuple, but I thought I'd bring it up.

stsievert replied:
> force the user to provide a tuple,

I like that better. That's I'm using in example code.

mrocklin commented on dask/array/rechunk.py:
```diff
@@ -220,9 +226,20 @@ def rechunk(x, chunks="auto", threshold=None, block_size_limit=None):
+            n_chunks = (n_chunks,) * x.ndim
+
+        chunksizes = [
+            _even_chunksize(s, n_chunk) if n_chunk not in ("auto", -1) else s
```
You might want to try `isinstance(n_chunk, str)` to support inputs like `(10, "100 MiB")`, which are allowed in other parts of this API.

mrocklin commented on dask/array/rechunk.py:
```diff
@@ -662,3 +680,86 @@ def format_plan(plan):
+    >>> y.chunks
+    ((296, 296, 296, 296, 296, 296, 296, 296, 296, 296, 296, 296, 296, 296, 296, 296, 294),)
+    >>> len(y.chunks[0])
+    17
```
I haven't gone through this logic yet, but a lot of it feels like logic that is already in `rechunk`.  I wonder if you could get away with calling something like the following:

```python
In [5]: x.rechunk(x.shape[0] // 16).chunks                                                                                                                                                                        
Out[5]: 
((314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  314,
  6),)
```

I guess you don't like the last 6 there though?

stsievert replied:
> I guess you don't like the last 6 there though?

Exactly. For Hyperband, each `partial_fit` call is assumed to receive the same amount of data, and `partial_fit` is called on each chunk. In this case, that means  scores after processing 6 and 314 examples are treated the same.

jsignell commented on dask/array/rechunk.py:
```diff
@@ -200,6 +201,13 @@ def rechunk(x, chunks="auto", threshold=None, block_size_limit=None):
+        ``chunks``; examples "auto" and "100MB".
+
+        If ``x`` has a single dimension and a int/float/string is passed, it's
+        case to a tuple.
```
```suggestion
        cast to a tuple.
```

jsignell commented on dask/array/rechunk.py:
```diff
@@ -662,3 +691,86 @@ def format_plan(plan):
+    Examples
+    --------
+    >>> import numpy as np
+    >>> import dask.array as da
```
I don't think this works as written. You are missing an import:

```python
from dask.array.rechunk import _even_chunksize
```
Also I get a different result: 

```python
((315,
  315,
  315,
  315,
  315,
  315,
  315,
  315,
  315,
  315,
  315,
  315,
  315,
  315,
  315,
  305),)
```

mrocklin commented:
It seems like there are two policies for how to split:

1.  Use a provided chunksize to start, and then leave any extra dangling off the end, like `(100, 100, 20)`
2.  Try to get even chunk sizes like `(110, 110)`

And there are two ways to specify roughly the size of chunk that we want

1.  A size in bytes, like "10 MiB"
2.  A number of chunks, like `100`

It feels slightly unclean to correlate these two "Oh hey, if you want even chunk sizes then make sure to use the `n_chunks` keyword, but if you want to have rigid chunk sizes with some extra overlap, then specify a number of bytes".  These two things seem unrelated to me.  I think that instead we should consider how to handle this more generally.  It would be good to get some other array computing folks in on this conversation (@rabernat @jakirkham  maybe?)

@stsievert if you're in a rush with this you might also consider pushing this into dask-ml, where we don't have to be quite as rigorous as in dask-array.

jakirkham commented:
What happens if one encounters a number that is just not evenly divisible (like `7`)?

stsievert commented:
> It feels slightly unclean to correlate these two [chunk size and if chunk size is even across chunks]

Would replacing `n_chunks` with a flag to balance the sizes of the chunks resolve this issue?

``` python
>>> x = np.arange(220)
>>> y = da.from_array(x).rechunk(chunks=100, approx=True)
>>> y.chunks
(110, 110)
```

This would also resolve my issue in the Hyperband use case.

> What happens if one encounters a number that is just not evenly divisible (like 7)?

You mean if the length of the Dask Array is 7? I just pushed a change that will do the following:

``` 
>>> y = da.from_array(np.arange(7)).rechunk(n_chunks=2)
>>> y.chunks
((4, 3),)
```

[Larger prime numbers][1] like 2129 have this behavior:

```
>>> da.from_array(np.arange(2129)).rechunk(n_chunks=7).chunks
((305, 305, 305, 305, 305, 305, 299),)
```

[1]:https://en.wikipedia.org/wiki/List_of_prime_numbers#The_first_1000_prime_numbers

stsievert commented:
I've pushed an example of what I think @mrocklin means in https://github.com/dask/dask/pull/6420#issuecomment-661247275. It's still rough. But now this test passes:

``` python
def test_balance_basics():
    x = da.from_array(np.arange(220))
    balanced = x.rechunk(chunks=100, balance=True)
    unbalanced = x.rechunk(chunks=100, balance=False)
    assert unbalanced.chunks[0] == (100, 100, 20)
    assert balanced.chunks[0] == (110, 110)
```

mrocklin commented:
I haven't looked at this yet, but in general I like the idea of a `balance=` keyword. This is on my TODO list for tomorrow.  Sorry for the delay.

mrocklin commented on dask/array/rechunk.py:
```diff
@@ -234,11 +248,20 @@ def rechunk(x, chunks="auto", threshold=None, block_size_limit=None):
         raise ValueError("Provided chunks are not consistent with shape")
+
+    if balance:
+        ideal_chunks = tuple(int(np.median(chunk)) for chunk in chunks)
+        even_chunks = tuple(
+            _even_chunksize(N, N // c) for N, c in zip(x.shape, ideal_chunks)
+        )
```
There is a lot of logic here with median and with the `_even_chunksize` function.  Why is this necessary?  Would it be ok to look at the shape, the number of chunks, and then do the logic in your `_get_chunks` function?  Functions with lots of tricky logic in them make me nervous.  They often cause error reports in the future.  Is there something about the choices that this code makes that are important?

stsievert replied:
Good point. The logic is convoluted. I think it's because I went from `n_chunks` to `balance`.

I think I see a way to simplify the logic in this PR. I'll check in after implementation.

mrocklin commented on dask/array/rechunk.py:
```diff
@@ -219,10 +222,21 @@ def rechunk(x, chunks="auto", threshold=None, block_size_limit=None):
+    ((167, 167, 167, 167, 167, 165), (1000,))
+
+    This means there isn't a chunk of size 4, which is what would happen
+    with ``x.rechunk(chunks=(1000 // 6, -1))``.
```
I modified this text.  I hope that you don't mind.

stsievert replied:
Yeah, that's fine. Thanks!

stsievert commented:
I've simplified the implementation significantly. The tests still pass without modification. Now, I have this docstring:

> balance : bool, default=False
> If True, try to make each chunk to be the same size.
>
> Reduce the number of chunks by one if there will be any small leftover chunks.  This allows using ``x.rechunk(chunks=len(x) // N, balance=True)`` to split into ``N`` chunks even if there are leftover small chunks.

I think I could make this change because I specify a threshold for "small chunk sizes" and I have modified the signatures for my new function `_balance_chunksizes`. Now the input and output are a tuple of chunk sizes.

Thanks for the suggestion @mrocklin; now the code is much simpler.

jsignell commented:
It seems like this is ready for a merge. Can you confirm @stsievert? @mrocklin in case you have any last comments.

stsievert commented:
It's definitely ready for another review. From my POV, it's also ready for merge.

jsignell commented:
> From my POV, it's also ready for merge.

I agree. I don't really see what the harm is in merging it since it's an isolated change.

jsignell commented:
I'm going to merge this unless there are any dissenters.
